### PR TITLE
Add offline use case for 'migrateFromSessionToken'

### LINKED ISSUE #2396: Graceful handling of the offline case during a sessionToken migration
Device could fail to perform a session token migration due to network issue; as discussed with @rfk, we should be fine to just "pretend it succeeded" at the api level, store the sessionToken internally in the account state, and attempt to recover organically over time.

Looking through A-C code, it seems like this approach will "just work".

cc @vladikoff

┆Issue is synchronized with this [Jira Story](https://jira.mozilla.com/browse/SYNC-534)
┆Story Points: 3
┆Epic: Fenix Migration
┆Sprint: SYNC - end 2020-01-31

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #2396

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
  - `swiftformat --swiftversion 4 megazords components/*/ios && swiftlint` runs without emitting any warnings or producing changes
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by vladikoff** [comment]
@eoger @rfk simpler approach to this now based on what we learned in the past PR. r?

**Comment by rfk** [comment]
Thanks @vladikoff; I'm adding @grigoryk as reviewer for the proposed approach from the consumer side.

File: CHANGES_UNRELEASED.md [hunk 1]
```diff
@@ -15,3 +15,16 @@
...
+
+### Breaking Changes
+
+* `migrateFromSessionToken` now returns a metrics JSON object if the migration succeeded.
+  ([#2492](https://github.com/mozilla/application-services/pull/2492))
+
+### Features
+
+* `migrateFromSessionToken` now handles offline use cases. It caches the data the consumers originally provide.
+  If there's no network connectivity then the migration could be retried using the new `retryMigrateFromSessionToken` method.
```
Comment (rfk) [inline CHANGES_UNRELEASED.md]: It's not clear from the description here exactly *how* the consumer should know whether or not to call this method (assuming that the migration could fail for some reasons that can be retried, and some reasons that cannot).

File: components/fxa-client/examples/migration.rs [hunk 2]
```diff
@@ -15,8 +14,22 @@ fn main() {
...
     let k_xcs: String = prompt_string("k_xcs").unwrap();
-    fxa.migrate_from_session_token(&session_token, &k_sync, &k_xcs, true)
-        .unwrap();
-    println!("WOW! You've been migrated.");
+    let migration_result =
+        match fxa.migrate_from_session_token(&session_token, &k_sync, &k_xcs, true) {
+            Ok(migration_result) => migration_result,
+            Err(err) => {
+                println!("Error: {}", err);
+                // test offline behaviour
```
Comment (rfk) [inline components/fxa-client/examples/migration.rs]: Unless this gets run in CI on a regular basis, I'm not sure we can claim it as a "test" ;-)

File: components/fxa-client/examples/migration.rs [hunk 3]
```diff
@@ -15,8 +14,22 @@ fn main() {
...
+                loop {
+                    thread::sleep(time::Duration::from_millis(5000));
+                    let retry = fxa.try_migration();
+                    match retry {
+                        Ok(result) => break result,
+                        Err(_) => println!("Retrying... Are you connected to the internet?"),
```
Comment (rfk) [inline components/fxa-client/examples/migration.rs]: Continuing discussion from [the previous PR](https://github.com/mozilla/application-services/pull/2481#pullrequestreview-344317428), each migration attempt could fail for either both fatal or non-fatal reasons. It would be useful for this example to show how to handle the two different cases. For example right now, if I enter an invalid sessionToken into this example script, then IIUC it will loop forever asking me whether I'm connected to the internet. We don't want to get anyone's migrated Fennec stuck in such a loop in practice.

Comment (rfk) [reply components/fxa-client/examples/migration.rs]: (I'll also note that "not connected to the internet" is not necessarily the only transient error; 500-level server errors should probably also be treated as transient and retried).

**Comment by vladikoff** [comment]
> automagical behaviour - a failed attempt persists FxA state, pretends to succeed, and will attempt to re-auth on FxA API access.

We tried that in the other approach and it didn't pan out. We will possibly revisit that in the future once we simplify the state persistence

**Comment by rfk** [comment]
> We tried that in the other approach and it didn't pan out

To add a bit of context from the previous conversations, that approach turned out to be surprisingly invasive, with lots of methods suddenly having the potential to modify and account state and need to trigger the persistence callback.

**Comment by vladikoff** [comment]
Updated!

File: components/fxa-client/examples/migration.rs (hunk 3 above)
Comment (rfk) [reply components/fxa-client/examples/migration.rs]: I know it's only for testing purposes, but I'd like to see this example use the same logic as we expect consumers to use in practice, which IIUC is basically:

```
match migrate_from_session_token(...) {
  Ok(result) => result,
  Err(...) => {
    while is_in_migration_state() {
       sleep(...)
       match retry_migrate_from_session_token() {
         Ok(result) => result,
         Err(...) => continue,
        }
    }
  }
}
```

File: CHANGES_UNRELEASED.md [hunk 4]
```diff
@@ -3,3 +3,17 @@
...
+
+* `migrateFromSessionToken` now returns a metrics JSON object if the migration succeeded.
+  ([#2492](https://github.com/mozilla/application-services/pull/2492))
+
+### Features
+
+* `migrateFromSessionToken` now handles offline use cases. It caches the data the consumers originally provide.
+  If there's no network connectivity then the migration could be retried using the new `retryMigrateFromSessionToken` method. 
+  Consumers may also use the `isInMigrationState` method to check if there's a migration in progress.
+  ([#2492](https://github.com/mozilla/application-services/pull/2492))
```
Comment (rfk) [inline CHANGES_UNRELEASED.md]: I suggest rewording this for clarity, to focus on the things that consumers need to know in the order they need to know them. Along the lines of:

* `migrateFromSessionToken` now has special handling for transient failures such as network errors. If the migration fails due to a transient error, then the provided credentials will be cached so that the migration can be retried later. Consumers can call the `isInMigrationState` method to check if there is a cached migration in progress, and `retryMigrateFromSessionToken` to retry it.

File: components/fxa-client/src/migrator.rs [hunk 5]
```diff
@@ -24,20 +31,44 @@ impl FirefoxAccount {
...
+            k_sync: k_sync.to_string(),
+            k_xcs: k_xcs.to_string(),
+            copy_session_token,
+            session_token: session_token.to_string(),
+        });
+
+        self.try_migration()
+    }
+
+    /// Check if the client is
```
Comment (rfk) [inline components/fxa-client/src/migrator.rs]: If the client is what? :-P

File: components/fxa-client/src/migrator.rs [hunk 6]
```diff
@@ -24,20 +31,44 @@ impl FirefoxAccount {
...
+    pub fn is_in_migration_state(&self) -> bool {
+        self.state.in_flight_migration.is_some()
+    }
+
+    pub fn try_migration(&mut self) -> Result<FxAMigrationResult> {
```
Comment (rfk) [inline components/fxa-client/src/migrator.rs]: Naming nit: I think the overall structure of the code would be a bit clearer if this were called `retry_migrate_from_session_token` for symmetry with what it's called in the higher layers.
